keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -613,7 +620,143 @@ sequence that it will not successfully decode.
+
+In cases 1-4, 7-8, and 9-10, if the marshalling side populates Raw with a non-nil slice, it is
+responsible for ensuring that that encoding of the slice contents matches the encoding that will be
+used to serialize the object containing the `RawExtension`. This is trivially ensured in cases 13
+and 14 because Protobuf is capable of representing `RawExtension` values containing arbitrary
+bytes. Protobuf is not a supported encoding for Unstructured objects. Existing programs have in
+practice stored JSON in the Raw field of `RawExtension`.
+
+#### Compatibility
+
+If the `RawExtension` marshalling and unmarshalling behavior for CBOR were to be implemented in
+exactly the same way as the existing JSON behaviors, the assumptions in many existing programs that
+the Raw field can be assigned to a slice of JSON bytes, or that the Raw bytes of an unmarshalled
+`RawExtension` are valid JSON, would be broken.
+
+The simple approach of automatically transcoding JSON to CBOR during CBOR marshalling, and
+transcoding CBOR to JSON during CBOR unmarshalling, would avoid breaking existing programs. However,
+the expense of transcoding to or from JSON would negate any performance advantage of a binary
+encoding. This expense would not be limited to a few API types: significant examples include the use
+of a `RawExtension` field in `metav1.WatchEvent` to represent each watch event's object state, or
+the arbitrary objects embedded in `admissionv1.AdmissionRequest`.
+
+Instead, automatic transcoding to and from JSON will become optional and initially enabled by
+default.
+
+For marshalling:
+
+1. If both Raw and Object are `nil`, `null` is returned.
+1. If Object is not `nil`, return the result of marshalling Object.
+1. If Raw is prefixed by the CBOR encoding of the head of the "self-described CBOR" tag, then return
+   the Raw bytes verbatim. This is the same prefix used by the CBOR `recognizer.RecognizingDecoder`
+   implementation, and all objects encoded by the CBOR `runtime.Encoder` implementation have this
+   prefix.
+1. Otherwise, assume Raw contains JSON and return the result of transcoding Raw from JSON to CBOR.
+
+Unmarshalling will behave the same for CBOR as it does for JSON and the input bytes will be copied
+verbatim to the Raw field. A Decoder implementation will be provided that delegates to another
+Decoder, then post-processes the result by transcoding `RawExtension` bytes to JSON before
+returning. The CBOR serializer available through `serializer.CodecFactory` will be wired to use
+this, allowing existing programs to continue to assume that unmarshalled Raw bytes contain JSON. The
+stream serializer will not. In practice, the watch decoder assumes that the non-stream serializer
+can directly decode the Raw bytes of a `metav1.WatchEvent` decoded by the stream serializer.
+
+TODO: What is the migration path? Feature gate to disable automatic transcoding on CBOR serializers in CodecFactory, guidance to use UniversalDeserializer, timeline...
```

Comment:
perhaps writing this in terms of client program expectations.

1. Proposed state for GA for naive clients: 
    1. client expects RawExtension to be JSON
    2. client receives RawExtension bytes as CBOR
    3. during decoding, the RawExtension is converted to JSON (this is expensive, but we do it to be compatible)
    4. client continues to parse the input as JSON
2. Proposed state for GA for advanced clients:
    1. client tolerates RawExtension to be JSON or CBOR
    2. client receives RawExtension bytes as CBOR
    3. during decoding, the RawExtension is left as CBOR (this is efficient)
    4. client detects RawExtension and parses either way (apimachinery should provide a very easy one-liner to do this and get an Unstructured out)
3. GA + default as CBOR for a year:
    1. Defaults switch with control via featuregate env var.
4. GA + default as CBOR for two year:
    1. featuregate control is removed, only control via code is now allowed.  This is because all existent clusters will be using CBOR and to end up broken the binaries had to update client libraries and test.

Reply:
Thanks, this is exactly what I needed.

> (apimachinery should provide a very easy one-liner to do this and get an Unstructured out)

I mentioned to @jpbetz that I would like to implement `UnstructuredConverter` on `RawExtension`, that might suffice?

Reply:
> I mentioned to @jpbetz that I would like to implement `UnstructuredConverter` on `RawExtension`, that might suffice?

Yes.

Reply:
Thanks. Incorporated this migration sequence into the doc.